### PR TITLE
Update Record Factory to use `sql_type` if `type` is missing

### PR DESCRIPTION
**Pull request by Criptak**
This updates the Record Factory to use the ActiveRecord Column's
`sql_type` if `type` is missing/nil.  This makes MR more flexible in
which columns can be used.  If you are using a field in the database
that isn't support by ActiveRecord (i.e. `json` or `jsonb` for
ActiveRecord earlier than version 4), you can define a factory that
matches the `sql_type` and still use the factory.

This also updates the factory to avoid setting any fields that are not
supported.  This is the least suprising behavior as it does not force the
end user to define every columne type that is used, just those that are
important enough to generate data for.

### CODE REVIEW

File: lib/mr/factory/record_factory.rb [hunk 1]
```diff
@@ -45,7 +45,12 @@ def apply_args(record, args = nil)
     def column_args
       @columns ||= non_association_columns(@record_class)
       @columns.inject({}) do |a, column|
-        a.merge(column.name.to_s => MR::Factory.send(column.type))
+        begin
+          column_type = column.type || column.sql_type.to_sym
```
Comment (jcredding) [inline lib/mr/factory/record_factory.rb]: @Criptak - Don't worry about the `to_sym` on the end of the sql type. I believe it returns a string which will work with `send`. I'd rather not pay any performance of converting the string to a symbol in this case.

File: lib/mr/factory/record_factory.rb [hunk 2]
```diff
@@ -45,7 +45,12 @@ def apply_args(record, args = nil)
...
+          a.merge(column.name.to_s => MR::Factory.send(column_type))
+        rescue StandardError
```
Comment (jcredding) [inline lib/mr/factory/record_factory.rb]: @Criptak - I think this is a little aggressive and it might hide errors. I'd prefer to either change this to `NoMethodError` or switch to an if.

For the `NoMethodError` I'd want to isolate generating the column value:

```ruby
column_type  = column.type || column.sql_type
column_value = begin
  MR::Factory.send(column_type)
rescue NoMethodError
  nil
end
a.merge(column.name.to_s => column_value)
```

This still has the problem of hiding no method errors in the factory methods themselves. For MR we test those individually so its not the biggest deal.

Alternatively, using an if:

```ruby
column_type  = column.type || column.sql_type
if MR::Factory.respond_to?(column_type)
  a.merge(column.name.to_s => MR::Factory.send(column_type))
else
  a
end
```

This doesn't hide any errors but I'm guessing it (generally) performs worse because of the call to `respond_to?` and the `if` (it may not perform worse than when the other implementation raises an error but that isn't the common case).

/cc @kellyredding

**Comment by jcredding** [comment]
@Criptak - Thanks for this. I'm mostly good with this but I'd like to get @kellyredding input on the catching of errors.

File: lib/mr/factory/record_factory.rb (hunk 2 above)
Comment (kellyredding) [inline lib/mr/factory/record_factory.rb]: @Criptak @jcredding I'd prefer the `if` approach b/c exceptions really shouldn't do control flow.

**Comment by Criptak** [comment]
@kellyredding @jcredding updated with @jcredding's comments

**Comment by jcredding** [comment]
@Criptak - I'm good with this :boom:

**Comment by kellyredding** [comment]
@Criptak please update the commit message and PR desc to paragraph break before "This also updates
the factory...".  This is a distinct thought that should be in a sep paragraph.  Thanks.

File: test/system/factory/record_factory_tests.rb [hunk 3]
```diff
@@ -42,6 +42,29 @@ class BuildingRealRecordTests < SystemTests
       assert_false user_record.new_record?
     end
 
+    should "use the `sql_type` if there is no ActiveRecord `type`" do
```
Comment (kellyredding) [inline test/system/factory/record_factory_tests.rb]: @Criptak move these up to jux next to the "attributes set using..." test above please.

File: test/system/factory/record_factory_tests.rb [hunk 4]
```diff
@@ -42,6 +42,29 @@ class BuildingRealRecordTests < SystemTests
...
+      number_column = UserRecord.columns.detect{ |c| c.name == "number" }
+      Assert.stub(number_column, :type){ nil }
+      Assert.stub(number_column, :sql_type){ 'integer' }
```
Comment (kellyredding) [inline test/system/factory/record_factory_tests.rb]: @Criptak I'd change this stub to do something other than what `:type` is without the stubbing.  Less risk of a false positive I think.

File: test/system/factory/record_factory_tests.rb [hunk 5]
```diff
@@ -42,6 +42,29 @@ class BuildingRealRecordTests < SystemTests
...
+
+    should "avoid setting a field if either `type` or `sql_type` is invalid" do
+      number_column = UserRecord.columns.detect{ |c| c.name == "number" }
+      Assert.stub(number_column, :type){ Factory.string }
+
+      user_record = @factory_class.new(UserRecord).instance
+      assert_nil user_record.number
+
+      Assert.stub(number_column, :type){ nil }
+      Assert.stub(number_column, :sql_type){ Factory.string }
```
Comment (kellyredding) [inline test/system/factory/record_factory_tests.rb]: @Criptak let's add another scenario to this test where you stub both type and sql type to be nil and show that you get a nil back - to be more thorough.

**Comment by kellyredding** [comment]
@Criptak a few testing comments - I'll look again once addressed.

**Comment by Criptak** [comment]
@jcredding @kellyredding updated to address the testing comments

**Comment by kellyredding** [comment]
@Criptak update the commit msg and PR desc and I'll merge this.  Thanks man!
